### ISSUE TITLE
HMR (live reload, fast refresh) broke render in React

### ISSUE THREAD

**Issue by dartess** [TanStack/virtual#750]
### Describe the bug

After the hot reload, new list items are not rendered

### Your minimal, reproducible example

https://stackblitz.com/edit/tanstack-virtual-dc5j1h?file=src%2Fmain.tsx,src%2FRowVirtualizerFixed.tsx&preset=node

### Steps to reproduce

1. open repl
2. edit `console.log('foo');` with some different text
3. press cmd/ctrl+S for save and live reload
4. scroll list
5. see that list is empty

### Expected behavior

it should just work

### How often does this bug happen?

Every time

### Platform

any platform , any bundler — vite, wepack, rspack

### tanstack-virtual version

3.8.1

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by mr-devboy** [comment]
Same here

**Comment by hophiphip** [comment]
TLDR: `scrollElement` gets cleared in `_didMount` and that is why virtualizer stops working. 



The only methods that can change `scrollElement` are: `_willUpdate` and `_didMount`. 

1. On the initial run `_willUpdate` and `_didMount` are called in this order:
![2024-08-29-16:45:41-screenshot](https://github.com/user-attachments/assets/aefcd879-67d2-4a8d-bd2c-5c82f2fbb2d5)

2. After triggering HMR  `_willUpdate` and `_didMount` will be called in this order:
![2024-08-29-16:48:18-screenshot](https://github.com/user-attachments/assets/92114f6d-9da9-4522-84e6-88c9fff2cd30)

[2] - This means `scrollElement` gets cleared and set to `null` which "breaks" the virtualizer.


You can debug this in this CodeSandbox: https://codesandbox.io/p/devbox/4wnzrm
You can also find fixed `useVirtualizerBase` in `src/hooks/useVirtualizerFixed.tsx`.


NOTE!: Im not 100% sure if this is a valid fix (but is seems to work in DEV and PROD, and tests pass), but I had to change 
`useVirtualizerBase` hook [L45-L47](https://github.com/TanStack/virtual/blob/90ede1e03c0314477056c99e6d646757ed2403d5/packages/react-virtual/src/index.tsx#L45) from:
```ts
React.useEffect(() => {
    return instance._didMount();
  }, []);
```

to:

```ts
React.useEffect(() => {
    return () => {
      instance._didMount();
    };
  }, []);
```
